crates/bevy_ecs/src/query/mod.rs

```diff
@@ -55,131 +55,6 @@ mod tests {
         assert_eq!(values, vec![&B(3)]);
     }
 
-    #[test]
-    fn query_filtered_exactsizeiterator_len() {
-        fn choose(n: usize, k: usize) -> usize {
-            if n == 0 || k == 0 || n < k {
-                return 0;
-            }
-            let ks = 1..=k;
-            let ns = (n - k + 1..=n).rev();
-            ks.zip(ns).fold(1, |acc, (k, n)| acc * n / k)
-        }
-        fn assert_combination<Q, F, const K: usize>(world: &mut World, expected_size: usize)
-        where
-            Q: WorldQuery,
-            F: WorldQuery,
-            F::ReadOnly: ArchetypeFilter,
-            for<'w> QueryFetch<'w, Q::ReadOnly>: Clone,
-            for<'w> QueryFetch<'w, F::ReadOnly>: Clone,
-        {
-            let mut query = world.query_filtered::<Q, F>();
-            let iter = query.iter_combinations::<K>(world);
-            let query_type = type_name::<QueryCombinationIter<Q, F, K>>();
-            assert_all_sizes_iterator_equal(iter, expected_size, query_type);
-        }
-        fn assert_all_sizes_equal<Q, F>(world: &mut World, expected_size: usize)
-        where
-            Q: WorldQuery,
-            F: WorldQuery,
-            F::ReadOnly: ArchetypeFilter,
-            for<'w> QueryFetch<'w, Q::ReadOnly>: Clone,
-            for<'w> QueryFetch<'w, F::ReadOnly>: Clone,
-        {
-            let mut query = world.query_filtered::<Q, F>();
-            let iter = query.iter(world);
-            let query_type = type_name::<QueryState<Q, F>>();
-            assert_all_sizes_iterator_equal(iter, expected_size, query_type);
-
-            let expected = expected_size;
-            assert_combination::<Q, F, 0>(world, choose(expected, 0));
-            assert_combination::<Q, F, 1>(world, choose(expected, 1));
-            assert_combination::<Q, F, 2>(world, choose(expected, 2));
-            assert_combination::<Q, F, 5>(world, choose(expected, 5));
-            assert_combination::<Q, F, 43>(world, choose(expected, 43));
-            assert_combination::<Q, F, 128>(world, choose(expected, 128));
-        }
-        fn assert_all_sizes_iterator_equal(
-            iterator: impl ExactSizeIterator,
-            expected_size: usize,
-            query_type: &'static str,
-        ) {
-            let size_hint_0 = iterator.size_hint().0;
-            let size_hint_1 = iterator.size_hint().1;
-            let len = iterator.len();
-            // `count` tests that not only it is the expected value, but also
-            // the value is accurate to what the query returns.
-            let count = iterator.count();
-            // This will show up when one of the asserts in this function fails
-            println!(
-                r#"query declared sizes:
-for query: {query_type}
-expected:      {expected_size}
-len():         {len}
-size_hint().0: {size_hint_0}
-size_hint().1: {size_hint_1:?}
-count():       {count}"#
```

Comment:
The test doesn't only check for `len` being consistent, but also `size_hint` being equivalent to `count`. Removing the `ExactSize` impl doesn't imply the `size_hint` cannot be accurate.

Note that this test both checks both the `QueryIter` and `QueryCombinationIter` counts. I suggest extracting the assert_combination, assert_all_sizes_equal, assert_all_sizes_iterator_equal functions, split this test in two, and not test `len` for `QueryCombinationIter`.

Also if you are running tests, I recommend you use `cargo test --package bevy_ecs query_filtered_exactsizeiterator_len` (or whatever new name you come up with) this will save you time.